compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -666,10 +682,7 @@ mkConDeclField env fieldName fieldTy = do
 isConstraint :: LF.Type -> Bool
 isConstraint = \case
     LF.TSynApp _ _ -> True
-    LF.TStruct fields -> and
-        [ isSuperClassField fieldName && isConstraint fieldType
-        | (fieldName, fieldType) <- fields
-        ]
+    LF.TStruct _ -> True
```

Comment:
(read the other answer first, this doesn't make sense otherwise)

A side effect of using expandTypeSynonyms below is that the instantiated method types will also be expanded. So the type synonyms (i.e. constraints) in the method types will also be expanded to their underlying TStruct definitions. To pick these up as constraints we modify this check.

It's simplest to say that any TStruct function argument we find is acting as a constraint, since that's true in the LF code we generate & consume.

Reply:
To avoid this, I could do the substitution below myself, instead of relying on expandTypeSynonyms. I think it's worth exploring if it makes the code simpler.

Reply:
I’ll leave it up to you if you want to explore it.

Reply:
I reverted the change to isConstraint by being more explicit about what I want expanded below (by using  `expandSynApp`, which only expands the top-level type synonym). :-)

This also gives me an idea for how to make data-dependencies faster... 🤔

Reply:
Not that worried about data-deps performance atm tbh given that we now don’t do this on every `daml build` anymore.

Reply:
Ok! If we ever need it, the idea is to do a structural comparison of typeclasses, in a "least fixed point"/graph-style computation, instead of using expandTypeSynonyms and alpha equivalence, when calculating which typeclasses are safe to reexport.